crates/bevy_picking/src/mesh_picking/ray_cast/intersections.rs

```diff
@@ -0,0 +1,405 @@
+    );
+
+    if let Some(indices) = indices {
+        // Make sure this chunk has 3 vertices to avoid a panic.
+        if indices.len() % 3 != 0 {
+            warn!("Index list not a multiple of 3");
+            return None;
+        }
+
+        // Now that we're in the vector of vertex indices, we want to look at the vertex
+        // positions for each triangle, so we'll take indices in chunks of three, where each
+        // chunk of three indices are references to the three vertices of a triangle.
+        for index in indices.chunks(3) {
+            let triangle_index = Some(index[0].into_usize());
+            let tri_vertex_positions = [
+                Vec3A::from(vertex_positions[index[0].into_usize()]),
+                Vec3A::from(vertex_positions[index[1].into_usize()]),
+                Vec3A::from(vertex_positions[index[2].into_usize()]),
+            ];
+            let tri_normals = vertex_normals.map(|normals| {
+                [
+                    Vec3A::from(normals[index[0].into_usize()]),
+                    Vec3A::from(normals[index[1].into_usize()]),
+                    Vec3A::from(normals[index[2].into_usize()]),
+                ]
+            });
+
+            let Some(hit) = triangle_intersection(
+                tri_vertex_positions,
+                tri_normals,
+                closest_hit_distance,
+                &mesh_space_ray,
+                backface_culling,
+            ) else {
+                continue;
+            };
+
+            closest_hit = Some(RayMeshHit {
+                point: mesh_transform.transform_point3(hit.point),
+                normal: mesh_transform.transform_vector3(hit.normal),
+                barycentric_coords: hit.barycentric_coords,
+                distance: mesh_transform
+                    .transform_vector3(mesh_space_ray.direction * hit.distance)
+                    .length(),
+                triangle: hit.triangle.map(|tri| {
+                    [
+                        mesh_transform.transform_point3a(tri[0]),
+                        mesh_transform.transform_point3a(tri[1]),
+                        mesh_transform.transform_point3a(tri[2]),
+                    ]
+                }),
+                triangle_index,
+            });
+            closest_hit_distance = hit.distance;
+        }
+    } else {
+        for i in (0..vertex_positions.len()).step_by(3) {
+            let triangle_index = Some(i);
+            let tri_vertex_positions = [
+                Vec3A::from(vertex_positions[i]),
+                Vec3A::from(vertex_positions[i + 1]),
+                Vec3A::from(vertex_positions[i + 2]),
+            ];
+            let tri_normals = vertex_normals.map(|normals| {
+                [
+                    Vec3A::from(normals[i]),
+                    Vec3A::from(normals[i + 1]),
+                    Vec3A::from(normals[i + 2]),
+                ]
+            });
```

Comment:
Would prefer this use `.chunks_exact(3).enumerate()` for the same reason as above.

Also, there's no guarantee there will be normals at the given index, and it will panic if they aren't there, so i'd recommend this code not use raw indexing.

Reply:
Using `get` for the normals is a 20-45% perf regression :/ I don't think I'll add that for now, if it ends up being a problem then we can reconsider. bevy_mod_picking and bevy_mod_raycast have been widely used for a long time, and I haven't seen anyone complain about panics.

Reply:
Ah that's annoying, do you remember which benchmark? I might have a play later.